torchinductor/triton_ops/autotune.py

```diff
@@ -1,70 +1,163 @@
 import builtins
+import copy
 import hashlib
 import json
 import logging
+import multiprocessing
 import os.path
-import time
+import threading
 from typing import List
 
+import torch
 import triton
 from triton import Config
 from triton import cdiv
 from triton import heuristics
 from triton import next_power_of_2
 from triton.ops.matmul import get_configs_io_bound
 from triton.ops.matmul_perf_model import early_config_prune as mm_early_config_prune
+from triton.runtime.jit import KernelInterface
+from triton.runtime.jit import get_cuda_stream
+from triton.testing import do_bench
 
 from torchinductor import config
 from torchinductor.ir import ReductionHint
 from torchinductor.triton_ops.mm_perf_model import estimate_matmul_time
 from torchinductor.utils import conditional_product
 
+from ..codecache import AsyncCompile
 from .conv_perf_model import early_config_prune as conv_early_config_prune
 from .conv_perf_model import estimate_conv_time
 
 log = logging.getLogger(__name__)
 
 
-class CachingAutotuner(triton.code_gen.Autotuner):
+class CachingAutotuner(KernelInterface):
     """
-    Simplified version of Triton autotuner that has no invalidation key
-    and caches the best config to disk.
+    Simplified version of Triton autotuner that has no invalidation
+    key and caches the best config to disk to improve cold start times.
+    Unlike the main triton Autotuner, this version can precompile all
+    configs, and does not rely on the Triton JIT.
     """
 
-    def __init__(self, kernel, arg_names, configs, save_cache_hook):
-        super().__init__(kernel, arg_names, configs, [], None, None)
-        assert not self.early_config_prune
-        assert not self.perf_model
+    def __init__(self, fn, meta, configs, save_cache_hook):
+        super().__init__()
+        self.fn = fn
+        self.meta = meta
         self.save_cache_hook = save_cache_hook
+        self.configs = configs
+        self.launchers = []
+        self.lock = threading.Lock()
+
+    def precompile(self):
+        with self.lock:
+            if self.launchers:
+                return
+            self.launchers = AsyncCompile.map(self._precompile_config, self.configs)
+            self.configs = None
+
+    def _precompile_config(self, cfg: triton.runtime.autotuner.Config):
+        """Ahead of time compile a given autotuner config."""
+        torch.cuda.set_device(torch.cuda.current_device())
+        compile_meta = copy.deepcopy(self.meta)
+        for k, v in cfg.kwargs.items():
+            compile_meta["constants"][self.fn.arg_names.index(k)] = v
+        compile_meta["num_warps"] = cfg.num_warps
+        compile_meta["num_stages"] = cfg.num_stages
+
+        if config.compile_threads > 1:
+            major, minor = torch.cuda.get_device_capability(compile_meta["device"])
+            compile_meta["cc"] = major * 10 + minor
```

Comment:
nit - if minor crosses double digits, which is valid semver, this will become wrong (1.11 becomes 21, so does 2.1) Consider just keeping semver (major.minor) or * 100 to reduce collision.

Reply:
This behavior is to match Triton's API.

Reply:
has never happened so far, and compute capabilities don't even follow semvers

Reply:
Point taken :P